File: Data Structures/Tree Traversal/Pre-Order Traversal/preorder.c
// Prg for tree traversals
#include <stdio.h>
#include <stdlib.h>
#include <malloc.h>

// AClass for a node
struct node
{
    int data;
    struct node *left;
    struct node *right;
};

// Allocation of New Node
struct node *newNode(int data)
{
    struct node *node = (struct node *)
        malloc(sizeof(struct node));
    node->data = data;
    node->left = NULL;
    node->right = NULL;

    return (node);
}


// PREORDER Traversal
void printPreorder(struct node *node)
{
    if (node == NULL)
        return;
    printf("%d ", node->data);  // print value
    printPreorder(node->left);  //recur on left child
    printPreorder(node->right); // recur on right child
}

// MAIN
int main()
{
    struct node *root = newNode(1);
    root->left = newNode(2);
    root->right = newNode(3);
    root->left->left = newNode(4);
    root->left->right = newNode(5);

    printf("\nPreorder traversal of binary tree is: \n");
    printPreorder(root);

    return 0;
}

File: Data Structures/Tree Traversal/Pre-Order Traversal/preorder.js
// Prg for tree traversals in JS

function Node(data) {
    this.data = data;
    this.left = null;
    this.right = null;
}

// Creating Nodes
var root = new Node(1);
var n1 = new Node(2);
var n2 = new Node(3);
var n3 = new Node(4);
var n4 = new Node(5);

// Setting Children
root.left = n1;
root.right = n2;
n1.left = n3;
n1.right = n4;


// PREORDER Traversal
function pre_order(root, nodes) {
    console.log(root.data);
    if (root && root.left) {
        pre_order(root.left, nodes);
    }
    if (root && root.right) {
        pre_order(root.right, nodes);
    }
    return nodes;
}


// MAIN
console.log("PreOrder Traversal: ")
pre_order(root, []);
